rfc/plugins.md

```diff
@@ -0,0 +1,235 @@
+# Plugins
+
+A plugin is an ember-cli addon that hooks into the ember-cli-deploy
+pipeline in order to add functionality to a deploy. Example addon
+functionality would be uploading assets to S3, writing the index.html to
+Redis, or notifying a Slack channel a deploy has completed.
+
+Plugins should focus on doing a specific task. Ember developers will
+compose and combine multiple plugins to fine tune their deployment
+process.
+
+### Identifying plugins
+
+Plugins are published to npm as ember-cli addons. They will also have
+the keyword ``ember-cli-deploy-plugin``.
+
+Any plugin that is installed will be loaded by default, similar to
+ember-cli addons.
+
+### Base plugin object
+
+Plugins will extend a base object provided by ember-cli-deploy. The API
+of this object will be small to start, but it will give us an easy place
+to give these plugins additional interfaces if needed.
+
+```javascript
+S3Plugin = DeployPlugin.extend({
+  ...
+});
+```
+
+### Hooks
+
+Plugins hook into the deployment pipeline with function names that match
+the steps in the pipeline process. These hooks will scoped under a
+special hooks object inside the plugin so that any private or plugin
+functions do not clash with any hook names that we add in
+the future.
+
+Although the hook functions are inside the hook object, they will be
+executed with a `this` binding that points to the plugin object.
+
+```javascript
+S3Plugin = {
+  hooks: {
+    assets: function() {
+      this.something();
+      this.uploadToS3();
+    })
+  },
+
+  something: function() { ... },
+
+  uploadToS3: function() { ... }
+}
+```
+
+### Provided properties
+
+Ember-cli-deploy will inject a number of properties onto each plugin.
+
+```javascript
+S3Plugin = {
+  ui: { ... },
+  appConfig: { ... },
+  deployConfig: { ... },
+  pluginConfig: { ... }
+}
+```
+
+Property | file | info
+--- | --- | ---
+`ui` | - | The ember-cli UI object that can be used to write to stdout.
+`appConfig` | `config/environment.js` | The application configuration.
+`deployConfig` | `config/deploy.js` | The deploy configuration file.
```

Comment:
Do we need `deployConfig` if we are also passing `pluginConfig` in?  Should a plugin care about any config other than it's own?  If the plugins are supposed to be independent entities then I'd imagine everything they need should be in their own config, no? What's an example of a use case where a plugin may need access to the rest of the stuff in `config/deploy.js`?